### PR TITLE
Compatibility with numpy 2

### PR DESCRIPTION
**Pull request by arunkannawadi**
This PR makes `ngmix` compatible with `numpy>=2` and `numpy<2`.

The recommended method is to replace `np.array(..., copy=False)` with `np.asarray` [link](https://numpy.org/devdocs/numpy_2_0_migration_guide.html#adapting-to-changes-in-the-copy-keyword). However, given the necessity of `ndmin` which is not available in `asarray`, I took the [scipy approach](https://github.com/scipy/scipy/pull/20172) here.

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -11,7 +11,7 @@ jobs:
     name: tests
     strategy:
       matrix:
-        pyver: ["3.9", "3.10", "3.11"]
+        pyver: ["3.9", "3.10", "3.11", "3.12"]
```
Comment (arunkannawadi) [inline .github/workflows/test.yml:14]: I added 3.12 to explicitly check this with the python version in the latest `rubin-env`. I left 3.9 as is, since it acts as an implicit check with `numpy<2`.

File: ngmix/defaults.py [hunk 2]
```diff
@@ -1,5 +1,10 @@
 import numpy as np
 
+# Shim to support numpy >= 2 and < 2.0.0.
+if np.lib.NumpyVersion(np.__version__) >= "2.0.0":
+    copy_if_needed = None
+else:
+    copy_if_needed = False
```
Comment (arunkannawadi) [inline ngmix/defaults.py:7]: `defaults` may not be the best place for this, but happy to move it wherever it might be more obvious for you.

File: ngmix/flags.py [hunk 3]
```diff
@@ -77,7 +79,7 @@ def get_flags_str(val, name_map=None):
     nstrs = []
     for pow in range(32):
         fval = 2**pow
-        if ((val & fval) != 0):
+        if ((np.uint32(val) & fval) != 0):
```
Comment (arunkannawadi) [inline ngmix/flags.py:82]: With `numpy>=2`, this now results in an OverflowError. We could either adopt limiting `pow` to `range(30)` as we did in [metadetect here](https://github.com/esheldon/metadetect/pull/224). I figured it's probably safer to cast it instead since that leaves the most-significant bits still usable.

Comment (beckermr) [reply ngmix/flags.py:82]: We should 

- limit the loop to 31
- raise a warning if the input value is bigger than 2**32-1 or < 0
- safely cast using `np.array(value).astype(dtype)`

A direct call to `np.uint32` is raising a warning that it will raise an error in the future for my local copy of numpy at 1.26.4.

File: .github/workflows/test.yml (hunk 1 above)
Comment (beckermr) [reply .github/workflows/test.yml:14]: SGTM!

File: ngmix/flags.py (hunk 3 above)
Comment (arunkannawadi) [reply ngmix/flags.py:82]: Oh, I didn't know there were going to be future releases of `numpy 1.26` that would convert this to an error.

Comment (beckermr) [reply ngmix/flags.py:82]: They might mean in numpy 2. IDK.

File: ngmix/flags.py [hunk 4]
```diff
@@ -74,10 +76,14 @@ def get_flags_str(val, name_map=None):
...
+    if val < 0 or val >= 2**32:
+        import warnings
+        warnings.warn(f"Flag value {val} is out of range [0, 2**32).")
+
     nstrs = []
-    for pow in range(32):
+    for pow in range(31):
         fval = 2**pow
-        if ((val & fval) != 0):
+        if ((np.array(val).astype(np.uint32) & fval) != 0):
```
Comment (beckermr) [inline ngmix/flags.py:86]: Let's move this cast up to the top and do it once.

Comment (arunkannawadi) [reply ngmix/flags.py:86]: I want to make sure we don't do too many redundant things. If we limit the loop to `range(31)`, then casting isn't even required. The error was arising from implicit casting of `fval` and not `val`. We should perhaps raise a `ValueError` if `val` is outside the specific range to avoid any underflow errors be done with it. For instance, if `val = np.int32(-5)`, we'd get a completely wrong result with just a seemingly benign warning.

Comment (arunkannawadi) [reply ngmix/flags.py:86]: My suggestion is to modify your list above to the following:

We should

- limit the loop to 31
- raise an exception if the input value is bigger than 2**32-1 or < 0

Comment (arunkannawadi) [reply ngmix/flags.py:86]: OK, in that case, having an upper limit doesn't make sense then. We raise an exception if `val` is negative and cast it to `uint64` safely always.

Comment (arunkannawadi) [reply ngmix/flags.py]: Well, what do you know. The code will now work even if `val` is `np.int8` etc.

Comment (arunkannawadi) [reply ngmix/flags.py]: My plan of always casting to `uint64` doesn't work for `numpy<2`. However, given the range of values that `fval` takes, which is hard coded in this routine,`uint32` should be sufficient. It does not pose any restriction on the exact integer type of `val` .

Comment (arunkannawadi) [reply ngmix/flags.py]: I'd hate to break anything because I didn't consider a corner case, so I added a small unit test as well.

File: ngmix/tests/test_flags.py [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+import pytest
+from ..flags import get_flags_str, NAME_MAP, LOW_DET
+
+
+@pytest.mark.parametrize(
+    "dtype", [np.int8, np.uint8, np.int16, np.uint16, np.int32, np.uint32, np.int64, np.uint64]
+)
+def test_get_flags_str(dtype):
+    """Test that get_flags_str works for different integer types."""
+    val = np.array(LOW_DET, dtype=dtype)
```
Comment (beckermr) [inline ngmix/tests/test_flags.py:13]: For the bigger types we should or with 2**45 to ensure things are OK.

Comment (arunkannawadi) [reply ngmix/tests/test_flags.py:14]: In `numpy<2`, bitwise or to `uint64` is prohibited for some reason. I'll condition on that if that's okay.

Comment (beckermr) [reply ngmix/tests/test_flags.py:14]: Ack I forgot and yes pls do!

File: ngmix/tests/test_flags.py [hunk 6]
```diff
@@ -11,8 +11,6 @@
 def test_get_flags_str(dtype):
     """Test that get_flags_str works for different integer types."""
     # Set a high-enough bit in addition to the lower bit to ensure things work.
-    val = np.array(LOW_DET, dtype=dtype)
-    if val.itemsize > 4 and dtype not in (np.uint64, ):
-        val |= 2**45
+    val = np.array(LOW_DET | 2**45).astype(dtype)
```
Comment (arunkannawadi) [inline ngmix/tests/test_flags.py:14]: Alright, this looks much cleaned than the test with all the conditions.

Comment (beckermr) [reply ngmix/tests/test_flags.py:14]: Yeah it just leaves the question of how numpy handles truncating things. That's ok I guess but maybe a comment would be good.

**Comment by beckermr** [comment]
@esheldon comments here or shall we merge?

**Review by esheldon — APPROVED** [review]

**Comment by esheldon** [comment]
Arun are you still making additions?

**Comment by arunkannawadi** [comment]
Nope, I don't have any more to add unless there are requests. Could you also please cut a release after merging this?

**Comment by arunkannawadi** [comment]
I shall, but I also notice that changes from #246 have not made it into a release yet. I will let you comment on that. This would bump up the minor version, right?

**Comment by esheldon** [comment]
I'll add notes for that.  I'll do the version bump when I make my changes.

**Comment by esheldon** [comment]
v2.4.0 is released

**Comment by arunkannawadi** [comment]
Thank you, Erin!
